### PR TITLE
Improve 1.21 client enchantments in legacy servers

### LINKED ISSUE #4235: Efficiency is too easy to desync
### `/viaversion dump` Output

https://dump.viaversion.com/4811191c9ece2273c9d268581078a3937a930e2b8dfc21ef06d897ede6eeed45

### Console Error

_No response_

### Bug Description

Related to https://github.com/ViaVersion/ViaVersion/pull/4060/files#diff-bddbbce71dc4ad767a0aa0ae4037a037505e51083cfbd135297a0849090f6ae8

For older servers (1.8) given the lack of a `PlayerInventorySlotChangeEvent` efficiency desyncs way too easily.

### Steps to Reproduce

Efficiency is only applied if you switch what your held item slot is, meaning a bunch of ways of getting efficiency end up with your pickaxe doing no efficiency unless you switch selected slot.

Some ways i've stumbled across which were annoying:
- Picking up a pickaxe directly in your held slot
- Moving an item in your inventory to your held slot (it be by clicks, or clicking the hotbar key to swap the item in)
- If a command or plugin modifies the item in your held slot (eg: /replaceitem )

### Expected Behavior

Enchantments (mainly, efficiency) should promptly apply whenever the item changes, in any way.

On legacy you should either listen to a bunch of events (item pickup, item destroy, inventory click), or, maybe better, do this at protocol level: whenever an item inventory change packet is sent, if it's for the held slot, update the attributes alongside it.
Is there any reason this can't be done at item remap level? ie: if an item has efficiency send it as having an attribute modifier for mining speed to the client.

### Additional Server Info

_No response_

### Checklist

- [X] Via plugins are only running on **EITHER** the backend servers (e.g. Paper) **OR** the proxy (e.g. Velocity), **not on both**.
- [X] I have included a ViaVersion dump.
- [X] If applicable, I have included a paste (**not a screenshot**) of the error.
- [X] I have tried the latest build(s) from https://ci.viaversion.com/ and the issue still persists.

### PR DESCRIPTION
**Pull request by Pablete1234**
Fixes #4235 

Legacy servers (anything without modern paper PlayerInventorySlotChangeEvent) will attempt to sync enchantments 
 in a few occasions:
- punching blocks (efficiency)
- closing inventory (armor)
- the clientbound set container slot packet (armor)

From testing it seems to be working ok (1.8 server) and even if there's still some ways to desync it (eg: right clicking to equip armor) i think it's good enough, worst case scenario reopening the inventory will fix it.

### CODE REVIEW

**Review by kennytv — COMMENTED** [review]
Mostly nitpicks

File: bukkit/src/main/java/com/viaversion/viaversion/bukkit/listeners/v1_20_5to1_21/PlayerChangeItemListener.java [hunk 1]
```diff
@@ -55,35 +54,26 @@ public void onPlayerItemHeld(final PlayerItemHeldEvent event) {
         sendAttributeUpdate(player, item, Slot.HAND);
     }
 
-    void sendAttributeUpdate(final Player player, @Nullable final ItemStack item, final Slot slot) {
-        final UserConnection connection = Via.getAPI().getConnection(player.getUniqueId());
-        if (connection == null || !isOnPipe(player)) {
-            return;
-        }
+    protected EfficiencyAttributeStorage getEfficiencyStorage(final UserConnection connection) {
+        return isOnPipe(connection) ? connection.get(EfficiencyAttributeStorage.class) : null;
```
[the remark this answers is not in the document]
Comment (Pablete1234) [reply bukkit/src/main/java/com/viaversion/viaversion/bukkit/listeners/v1_20_5to1_21/PlayerChangeItemListener.java:58]: it effectively speaking acts as a nullcheck for UserConnection, which i think could be null

File: common/src/main/java/com/viaversion/viaversion/rewriter/ItemRewriter.java [hunk 2]
```diff
@@ -36,14 +36,14 @@
...
     T extends Protocol<C, ?, ?, S>> extends RewriterBase<T> implements com.viaversion.viaversion.api.rewriter.ItemRewriter<T> {
-    private final Type<Item> itemType;
-    private final Type<Item> mappedItemType;
-    private final Type<Item[]> itemArrayType;
-    private final Type<Item[]> mappedItemArrayType;
-    private final Type<Item> itemCostType;
-    private final Type<Item> mappedItemCostType;
-    private final Type<Item> optionalItemCostType;
-    private final Type<Item> mappedOptionalItemCostType;
+    protected final Type<Item> itemType;
```
Comment (FlorianMichael) [inline common/src/main/java/com/viaversion/viaversion/rewriter/ItemRewriter.java:39]: No need to make them protected, the important one have getter functions at the bottom which should be used.

File: common/src/main/java/com/viaversion/viaversion/protocols/v1_20_5to1_21/rewriter/BlockItemPacketRewriter1_21.java [hunk 3]
```diff
@@ -67,7 +77,32 @@ public void registerPackets() {
...
+            final Item item = handleItemToClient(wrapper.user(), wrapper.read(itemType));
+            wrapper.write(mappedItemType, item);
+
+            // When a players' armor is set, update their attributes
+            if (containerId != 0 || slotId > BOOTS_SLOT || slotId < HELMET_SLOT || slotId == CHESTPLATE_SLOT) {
+                return;
+            }
+
+            final EfficiencyAttributeStorage storage = wrapper.user().get(EfficiencyAttributeStorage.class);
+            if (storage == null) {
```
Comment (FlorianMichael) [inline common/src/main/java/com/viaversion/viaversion/protocols/v1_20_5to1_21/rewriter/BlockItemPacketRewriter1_21.java:93]: Can be removed, the storage is always present

**Review by kennytv — APPROVED** [review]
Thanks!

**Pull request merged**